### ISSUE TITLE
Reboot Hydro Platinum device when reads fail

### ISSUE THREAD

**Issue by EvanMulawski** [EvanMulawski/FanControl.CorsairLink#80]
From #76, #78, #127, #130

When a Hydro Platinum device read fails multiple times, iCUE reboots the device:

https://forum.corsair.com/forums/topic/176494-corsair-h100i-platinum-se-randomly-stops-and-restarts/
https://forum.corsair.com/forums/topic/171275-hydro-h100i-connectingdisconnecting/

Implement reboot logic when a read fails by sending the reboot packet.

Notes:

- Based on an observation in #78, the device accepts writes and adjusts pump/fan speeds accordingly even when this issue occurs (implies only reads fail).

**Comment by EvanMulawski** [comment]
@jflevesque @Profetorum @BrendanMan3 @larz101 Please try the attached build. This will attempt to reboot the Hydro Platinum device if reads start failing (just like iCUE does). If the device reboots, the latest CorsairLink.log file will have "Rebooting device" followed by a read of the device state. Thanks!

[FanControl.CorsairLink-HydroPlatinumReboot-v20231223r001.zip](https://github.com/EvanMulawski/FanControl.CorsairLink/files/13760164/FanControl.CorsairLink-HydroPlatinumReboot-v20231223r001.zip)

**Comment by BrendanMan3** [comment]
[CorsairLink.log](https://github.com/EvanMulawski/FanControl.CorsairLink/files/13760316/CorsairLink.log)

**Comment by EvanMulawski** [comment]
@BrendanMan3 Do you have any other software running that is communicating with the device (iCUE, Corsair Service, OpenRGB, SignalRGB, monitoring software, etc.)?

**Comment by BrendanMan3** [comment]
[CorsairLink.log](https://github.com/EvanMulawski/FanControl.CorsairLink/files/13760356/CorsairLink.log)
[log.txt](https://github.com/EvanMulawski/FanControl.CorsairLink/files/13760357/log.txt)
Corsair looks to all have been shut down. The logs I just sent were after running fine for a few minutes.

I was running signal RGB. I'm testing now with it disabled.

**Comment by BrendanMan3** [comment]
I can run SignalRGB with the H115i disabled and fan control has not (yet) thrown any errors.

**Comment by Profetorum** [comment]
> @jflevesque @Profetorum @BrendanMan3 @larz101 Please try the attached build. This will attempt to reboot the Hydro Platinum device if reads start failing (just like iCUE does). If the device reboots, the latest CorsairLink.log file will have "Rebooting device" followed by a read of the device state. Thanks!
> 
> [FanControl.CorsairLink-HydroPlatinumReboot-v20231223r001.zip](https://github.com/EvanMulawski/FanControl.CorsairLink/files/13760164/FanControl.CorsairLink-HydroPlatinumReboot-v20231223r001.zip)

[CorsairLink.log](https://github.com/EvanMulawski/FanControl.CorsairLink/files/13762578/CorsairLink.log)
Still erroring out. I think it actually rebooted the pump or something, since the pump lights changed color, but apparently it didn't do much

**Comment by EvanMulawski** [comment]
@BrendanMan3 What version of SignalRGB? Compatibility with Hydro Platinum devices requires SignalRGB `2.3.45-beta` or later.

**Comment by EvanMulawski** [comment]
@Profetorum Thanks. I think I need to close and reopen the device after rebooting - will have another build soon.

**Comment by EvanMulawski** [comment]
New build.

[FanControl.CorsairLink-HydroPlatinumReboot-v20231224r002.zip](https://github.com/EvanMulawski/FanControl.CorsairLink/files/13763603/FanControl.CorsairLink-HydroPlatinumReboot-v20231224r002.zip)

**Comment by BrendanMan3** [comment]
Errors detected using 24r002 build while the H115i is enabled in SignalRGB - Will test with it disabled.
[CorsairLink.log](https://github.com/EvanMulawski/FanControl.CorsairLink/files/13776202/CorsairLink.log)

**Comment by BrendanMan3** [comment]
With the H115i disabled in Signal RGB I have not had any issues after 6 hours of running.

**Comment by EvanMulawski** [comment]
New build. Still trying to successfully reboot the device.

[FanControl.CorsairLink-HydroPlatinumReboot-v20231227r001.zip](https://github.com/EvanMulawski/FanControl.CorsairLink/files/13781981/FanControl.CorsairLink-HydroPlatinumReboot-v20231227r001.zip)

**Comment by Profetorum** [comment]
Going to download this one as well. 
Haven't had a pump failure with the previous build, it never restarted. So I can't tell if it was working.

**Comment by chjohans** [comment]
I too get errors with my H115i when running SignalRGB, the latest beta v 2.3.47-beta. If I disable the H115i in SignalRGB everything works just fine. I'm running HWiNFO64 as well, and it too looses the H115i sensore readings when this happens. This makes me question if the CorsairLink mutex implementation for Corsair Pro/Platinium AiOs they supposedly implemented in v 2.3.45-beta is working as intended: https://docs.signalrgb.com/changelogs/2-3-45

So I don't think the issue is with this plugin.

I've not had much luck in communicating with SignalRGB support in the past (I'm only a "free" user).

If you @EvanMulawski has communicated with them re the CorsairLink Mutex in the past then maybe you have a better channel to give them feedback on this?

**Comment by EvanMulawski** [comment]
@chjohans Can you attach your latest CorsairLink.log file? I'm thinking it could just be an inherent issue with the AIO's microcontroller and/or firmware - which is why iCUE has logic built-in to reboot these devices when it can no longer read from them.

**Comment by EvanMulawski** [comment]
I just ordered an H150i RGB PRO XT to test and troubleshoot this locally - will provide a status update after it arrives.

**Comment by chjohans** [comment]
Wow, thanks a million @EvanMulawski for your commitment to this, which is nothing short of impressive. I'll buy you a cup of coffee or two again shortly, and attach my logs as soon as I return home in a few hours.

**Comment by Meh7991** [comment]
I just installed the latest release of this plugin, and updated SignalRGB to 2.3.47-beta to see if I would have any luck with my H150i RGB Pro XT but nope. 
SRGB constantly logs "CRC error from color packet", as it used to for me when I was having issues. 

Exiting FanControl stops those errors being logged in SRGB. 

I have set my lighting effects to have a dim white background colour, so as a workaround for the time being I have just set my H150i to a static dim white colour in iCUE. 

As you have ordered a the same cooler that I have, @EvanMulawski, I'll be willing to share logs or anything again if you need it.

**Comment by chjohans** [comment]
Running HWiNFO64 with "CorsairLink and Asetek Support" enabled and Fan Control with your plugin in parallel works just fine, so it seems like the shared mutex is doing its job when used correctly. I can see that HWiNFO sometimes loses Corsair readings for a very brief moment, probably because the mutex is locked during its update cycle, and that is not unexpected.

But, when I enable my "Corsair Hydro H115i RGB PRO XT" in SingnalRGB (v2.3.47-beta) then HWiNFO immediately loses all reading from this device, and FanControl starts to complain about errors as well. Please see the attached CorsairLink.log. Nothing is logged in log.txt

When this happens I have to reboot my computer to get my AiO back in HWiNFO and FanControl again. Just stopping SignalRGB and restarting Fan Control and HWiNFO makes no difference.

[CorsairLink.log](https://github.com/EvanMulawski/FanControl.CorsairLink/files/13805105/CorsairLink.log)

**Comment by EvanMulawski** [comment]
I was able to reproduce this after a few minutes of testing - currently investigating.

**Comment by EvanMulawski** [comment]
New build. This will successfully reboot the device when necessary to restore Fan Control and SignalRGB connectivity.

[FanControl.CorsairLink-HydroPlatinumReboot-v20240105r001.zip](https://github.com/EvanMulawski/FanControl.CorsairLink/files/13848514/FanControl.CorsairLink-HydroPlatinumReboot-v20240105r001.zip)

Testing notes:

- I ran a USB capture while both SignalRGB and Fan Control were running concurrently. It appears that version `2.3.47-beta+ef6c0587` does use the well-known mutex correctly. I also confirmed this by monitoring the mutex when only SignalRGB was running.

- SignalRGB reports about a 6fps loss while Fan Control is running and the frame times spike every second as it waits to reacquire the mutex. (I was seeing 21fps without Fan Control running and 15fps with both running.)

- With **only** SignalRGB running (i.e. Fan Control **not running concurrently**), the device still got into a bad state after a short period of time. (Launching Fan Control successfully rebooted the device and both applications recovered.)

---

@Heal-Bot An observation from the packet capture:

When SignalRGB acquires the mutex, it should read the state immediately. Byte index 1 will be the previous sequence number - use this to generate the next sequence numbers.

Additionally, you may want to consider ignoring the responses from all lighting and cooling commands - I have found these to be unreliable and they frequently fail CRC verification. The read operation itself isn't necessary and was removed from OpenRGB.

**Comment by Meh7991** [comment]
v20240105r001 doesn't seem any different for me. 

I'm running FanControl v177, and SignalRGB 2.3.47-beta+ef6c0587

Here are my logs (with FANCONTROL_CORSAIRLINK_DEBUG_LOGGING_ENABLED enabled) - ~10 minutes: 
[CorsairLink.zip](https://github.com/EvanMulawski/FanControl.CorsairLink/files/13849470/CorsairLink.zip)
The device started bugging out after around 2 minutes of being enabled. 

Here are my logs (with FANCONTROL_CORSAIRLINK_DEBUG_LOGGING_ENABLED disabled) - ~10 minutes: 
[CorsairLink.zip](https://github.com/EvanMulawski/FanControl.CorsairLink/files/13849602/CorsairLink.zip)
The device seemed more stable this time, lighting effects were still a bit laggy but it began bugging out after ~6 minutes of being enabled (instead of displaying a static blue colour as below, it started flickering in rainbow colours towards the centre). 

Here is a physical picture of my cooler, as you can see in my below screenshot I have the 'Fireworks' effect set - but this is what my cooler ends up doing eventually: 
![image](https://github.com/EvanMulawski/FanControl.CorsairLink/assets/76539753/36e800da-499c-466f-b728-3b527b57ee86)
That blue colour is the 'shutdown' effect I have set in SignalRGB, not sure if that has any relevance. 

Here is a screenshot of the device's log in SignalRGB: 
![image](https://github.com/EvanMulawski/FanControl.CorsairLink/assets/76539753/aa039400-9fd6-43f2-957d-97c987477464)

I didn't see the framerate change at all from 15fps. 

Let me know if you need any more information. Happy to test out different builds.

**Comment by chjohans** [comment]
The attached version worked a bit better. I could now start SignalRGN and enable my H115i without FanControl instantly showing error messages.

But after a few minutes of playing around with SignalRGB the same error came back, and just as before the only way to get FanControl to work again is a full reboot. If I just restart FanControl (after stopping SIgnalRGB) it will not initialize, eg it will not find any Corsair sensors.

Please see the attached log file. Do take note that it does not contain logging from the restart attempt after the error happened. If ytou need that as well please let me know.

[CorsairLink.15.log](https://github.com/EvanMulawski/FanControl.CorsairLink/files/13850270/CorsairLink.15.log)

**Comment by EvanMulawski** [comment]
@Meh7991 @chjohans Can you run this build without SignalRGB running and let me know if you encounter any issues? Also, please try running HWiNFO64 (with Corsair support enabled) and [this build of OpenRGB](https://github.com/EvanMulawski/FanControl.CorsairLink/issues/101#issuecomment-1806536056) concurrently. Thanks!

**Comment by Profetorum** [comment]
Sorry guys I might be wrong here, not at my PC right now, but iirc the hwinfo says it should be disabled in order to skip polling for Corsair devices? Can someone check

**Comment by EvanMulawski** [comment]
> Sorry guys I might be wrong here, not at my PC right now, but iirc the hwinfo says it should be disabled in order to skip polling for Corsair devices? Can someone check

Only if iCUE or other applications that don't use the Corsair mutex/lock are running.

**Comment by chjohans** [comment]
> @Meh7991 @chjohans Can you run this build without SignalRGB running and let me know if you encounter any issues? Also, please try running HWiNFO64 (with Corsair support enabled) and [this build of OpenRGB](https://github.com/EvanMulawski/FanControl.CorsairLink/issues/101#issuecomment-1806536056) concurrently. Thanks!

I'm always running HWiNFO (with Corsair support enabled) :=)

I have been running this build together with HWiNFO with no issues since my post above. So without SIgnalRGB running it seems just fine.

I have been running the build of OpenRGB you are linking too together with HWiNFO and your FC plugin before with no issues, but I'll do some more testing with that and this build of your plugin and report back.

**Comment by Meh7991** [comment]
> @Meh7991 @chjohans Can you run this build without SignalRGB running and let me know if you encounter any issues? Also, please try running HWiNFO64 (with Corsair support enabled) and [this build of OpenRGB](https://github.com/EvanMulawski/FanControl.CorsairLink/issues/101#issuecomment-1806536056) concurrently. Thanks!

Seems to be no issues for me running this setup.

**Comment by chjohans** [comment]
> > @Meh7991 @chjohans Can you run this build without SignalRGB running and let me know if you encounter any issues? Also, please try running HWiNFO64 (with Corsair support enabled) and [this build of OpenRGB](https://github.com/EvanMulawski/FanControl.CorsairLink/issues/101#issuecomment-1806536056) concurrently. Thanks!
> 
> I'm always running HWiNFO (with Corsair support enabled) :=)
> 
> I have been running this build together with HWiNFO with no issues since my post above. So without SIgnalRGB running it seems just fine.
> 
> I have been running the build of OpenRGB you are linking too together with HWiNFO and your FC plugin before with no issues, but I'll do some more testing with that and this build of your plugin and report back.

Been playing around with all kinds of effects in OpenRGB for the past hour or so, no issues at all. I was running HwiNFO and FanCOntrol at the same time of course.

**Comment by BrendanMan3** [comment]
Successfully rebooting the H115i when it fails and is enabled in SignalRGB.

`2024-01-11T17:05:36.7297470Z [WRN] Plugin: Refresh skipped - refresh already in progress.
2024-01-11T17:05:38.4946964Z [ERR] H115i Platinum (8E43FDD6F5CED65912D9D8CBD34D1A3C): An error occurred refreshing device 'H115i Platinum (8E43FDD6F5CED65912D9D8CBD34D1A3C)' (\\?\hid#vid_1b1c&pid_0c17#8&27c7b466&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030})

Type: CorsairLink.CorsairLinkDeviceException
Message: Communication failure.
Source: FanControl.CorsairLink
Stack Trace:
   at CorsairLink.Devices.HydroPlatinumDevice.SendCommand(Byte command, ReadOnlySpan`1 data)
   at CorsairLink.Devices.HydroPlatinumDevice.ReadState()
   at CorsairLink.Devices.HydroPlatinumDevice.<Refresh>g__RefreshImpl|39_0()
   at CorsairLink.Devices.HydroPlatinumDevice.Refresh()
   at FanControl.CorsairLink.CorsairLinkPlugin.<>c__DisplayClass41_0.<RefreshAsync>b__0()
Data:
    writeBuffer: 003F08FF00FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF32
    readBuffer: 0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
Inner Exception:
    Type: System.TimeoutException
    Message: Operation timed out.
    Source: HidSharp
    Stack Trace:
   at HidSharp.Platform.Windows.NativeMethods.OverlappedOperation(IntPtr ioHandle, IntPtr eventHandle, Int32 eventTimeout, IntPtr closeEventHandle, Boolean overlapResult, NativeOverlapped* overlapped, UInt32& bytesTransferred)
   at HidSharp.Platform.Windows.WinHidStream.Read(Byte[] buffer, Int32 offset, Int32 count)
   at CorsairLink.Devices.HydroPlatinumDevice.WriteAndRead(Byte[] writeBuffer, Byte[] readBuffer)
   at CorsairLink.Devices.HydroPlatinumDevice.SendCommand(Byte command, ReadOnlySpan`1 data)


2024-01-11T17:05:38.7586175Z [WRN] Plugin: Refresh skipped - refresh already in progress.`

**Comment by EvanMulawski** [comment]
New build.

[FanControl.CorsairLink-HydroPlatinumReboot-v20240112r001.zip](https://github.com/EvanMulawski/FanControl.CorsairLink/files/13914820/FanControl.CorsairLink-HydroPlatinumReboot-v20240112r001.zip)

**Comment by Meh7991** [comment]
> New build.
> 
> [FanControl.CorsairLink-HydroPlatinumReboot-v20240112r001.zip](https://github.com/EvanMulawski/FanControl.CorsairLink/files/13914820/FanControl.CorsairLink-HydroPlatinumReboot-v20240112r001.zip)

This build looked to be working perfectly with SignalRGB running, and the H150i enabled in it. I thought to myself, "omg finally, everything is synced!!". 

Five minutes or so later: 
![image](https://github.com/EvanMulawski/FanControl.CorsairLink/assets/76539753/a71030a7-0262-4e26-a3e5-110484bbeb6a)

Here is my log, might be a bit large because I forgot to delete it before running the new build: 
[CorsairLink.zip](https://github.com/EvanMulawski/FanControl.CorsairLink/files/13917543/CorsairLink.zip)

I had to fully power cycle my PC in order for my H150i to become recognisable again in FanControl. 

Looks to be getting close to being stable though! The lighting effects looked much smoother than with the v20240105r001 build.

**Comment by EvanMulawski** [comment]
I am making some changes to the SignalRGB Hydro Platinum plugin code and will post the modified version after testing (with instructions).

**Comment by EvanMulawski** [comment]
Updated SignalRGB plugin attached for SignalRGB `2.3.45-beta` and above.

[signalrgb_hydro_platinum_plugin_v20240112r001.zip](https://github.com/EvanMulawski/FanControl.CorsairLink/files/13926403/signalrgb_hydro_platinum_plugin_v20240112r001.zip)

## Instructions

1. Unblock the downloaded ZIP file. (Properties > Unblock > OK)
2. Exit SignalRGB.
3. Open the directory `C:\Users\<username>\AppData\Local\VortxEngine\app-<version>\Signal-x64\Plugins\Corsair\Platinum_Pro_XT_AIO`
4. Copy the existing `Platinum_Pro_XT_Elite_AIO.js` file to another location to create a backup (do not create a copy in this directory).
5. Replace the file with the attached file.
6. Launch SignalRGB.

**Comment by Meh7991** [comment]
> Updated SignalRGB plugin attached for SignalRGB `2.3.45-beta` and above.
> 
> [signalrgb_hydro_platinum_plugin_v20240112r001.zip](https://github.com/EvanMulawski/FanControl.CorsairLink/files/13926403/signalrgb_hydro_platinum_plugin_v20240112r001.zip)
> 
> ## Instructions
> 1. Unblock the downloaded ZIP file. (Properties > Unblock > OK)
> 2. Exit SignalRGB.
> 3. Open the directory `C:\Users\<username>\AppData\Local\VortxEngine\app-<version>\Signal-x64\Plugins\Corsair\Platinum_Pro_XT_AIO`
> 4. Copy the existing `Platinum_Pro_XT_Elite_AIO.js` file to another location to create a backup (do not create a copy in this directory).
> 5. Replace the file with the attached file.
> 6. Launch SignalRGB.

Using your CorsairLink plugin v20240112r001, SignalRGB v2.3.47-beta+ef6c0587, and your attached SignalRGB plugin, everything seems to be working perfectly! 

I will leave this running for the next several hours and report any findings.

**Comment by Meh7991** [comment]
After about 1 hour of running, FanControl popped up with an error alert. 
[CorsairLink.zip](https://github.com/EvanMulawski/FanControl.CorsairLink/files/13927792/CorsairLink.zip)

**Comment by chjohans** [comment]
My results were a bit different, no issues so far after running this version for about 8 hours with SignalRGB effects enabled all the time. SignaRGB v2.3.47-beta, the latest version of FanControl with this plugin, and HWiNFO with "CorsairLink and Asetek Support" enabled. No errors were observed and no errors in any log files so this seems to be working for me and my H115i.

**Comment by EvanMulawski** [comment]
> After about 1 hour of running, FanControl popped up with an error alert. [CorsairLink.zip](https://github.com/EvanMulawski/FanControl.CorsairLink/files/13927792/CorsairLink.zip)

Did the device stop functioning after those errors? The checksum errors are typically harmless but it is interesting that 10 occurred in a row.

---

> My results were a bit different, no issues so far after running this version for about 8 hours with SignalRGB effects enabled all the time. SignaRGB v2.3.47-beta, the latest version of FanControl with this plugin, and HWiNFO with "CorsairLink and Asetek Support" enabled. No errors were observed and no errors in any log files so this seems to be working for me and my H115i.

I experienced the same thing with my H150i PRO XT - no errors requiring a device reboot and only sporadic checksum errors in the log after 6 hours of testing.

**Comment by Meh7991** [comment]
> > After about 1 hour of running, FanControl popped up with an error alert. [CorsairLink.zip](https://github.com/EvanMulawski/FanControl.CorsairLink/files/13927792/CorsairLink.zip)
> 
> Did the device stop functioning after those errors? The checksum errors are typically harmless but it is interesting that 10 occurred in a row.

I will re-install these plugins and see if that was just a one-off glitch. 
When I saw that error pop up earlier, I basically immediately quit FanControl and SignalRGB and reverted to the working plugins, all has been fine ever since (apart from the lack of synced lighting of course haha). 
But, if I remember correctly, I swear I saw the fan speeds for the cooler say 0RPM in FC - could be wrong about that, will see if/when it errors again. 

I did not need to power cycle my system for FC to recognise the cooler/fans again, however.

**Comment by Meh7991** [comment]
> After about 1 hour of running, FanControl popped up with an error alert. [CorsairLink.zip](https://github.com/EvanMulawski/FanControl.CorsairLink/files/13927792/CorsairLink.zip)

Looks like this was a one off glitch. I have had the new plugins running for almost 10 hours straight now and there hasn't been any errors. I'll attach my log file, which also includes ~2 hours from yesterday evening. 
[CorsairLink.zip](https://github.com/EvanMulawski/FanControl.CorsairLink/files/13933246/CorsairLink.zip)

If I come across any errors or anything, I will be sure to update here, otherwise this seems as stable as it'll get! 
Good job with with everything, @EvanMulawski!

**Comment by chjohans** [comment]
My PC is "always on", but goes to sleep when not being used. Attaching the CorsairLink.log from the past 2+ days. I get some "Checksum Failure" in the logfile, but I have not gotten any error messages in the UI yet. I don't see any errors in SIngnalRGB either and it has been running with a constantly changing RGB effect the whole time.

[CorsairLink.log](https://github.com/EvanMulawski/FanControl.CorsairLink/files/13941457/CorsairLink.log)

**Comment by EvanMulawski** [comment]
Thanks for the updates! I'm going to reduce the log level of the checksum error to Warning and track those differently as the device is still performing as expected.

**Comment by chjohans** [comment]
@EvanMulawski After SignalRGB auto-updated to V2.3.48-beta the error is back, it was solved by replacing the file Platinum_Pro_XT_Elite_AIO.js as per your post further up.

Will this change somehow be included with SignalRGB at some point or do we have to replace this file on every update?

**Comment by EvanMulawski** [comment]
> @EvanMulawski After SignalRGB auto-updated to V2.3.48-beta the error is back, it was solved by replacing the file Platinum_Pro_XT_Elite_AIO.js as per your post further up.

I believe SignalRGB re-downloads all plugins when updating.

> Will this change somehow be included with SignalRGB at some point or do we have to replace this file on every update?

The SignalRGB team will need to implement the changes their official version of the plugin.

/cc @Heal-Bot

**Comment by chjohans** [comment]
> > @EvanMulawski After SignalRGB auto-updated to V2.3.48-beta the error is back, it was solved by replacing the file Platinum_Pro_XT_Elite_AIO.js as per your post further up.
> 
That I also figured out since the file had changed with the update, and re-placing it again solved the issue.

> I believe SignalRGB re-downloads all plugins when updating.
> 
> > Will this change somehow be included with SignalRGB at some point or do we have to replace this file on every update?
> 
> The SignalRGB team will need to implement the changes their official version of the plugin.
> 
I also figured out that much, I guess I should have been crystal clear in my question. What I really wonder is where the changed file comes from, is it something you just did as a quick-fix to get this to work, or has it been done by the SignalRGB team, or both? And are someone communicating these needed changes to their dev team?

I am guessing that this (bug?) in the Platinum_Pro_XT_Elite_AIO.js also affect other apps (besides FanControl) ability to run in parallel with SIgnalRGB even when they have implemented the shared mutex correctly?

> /cc @Heal-Bot

**Comment by EvanMulawski** [comment]
@Heal-Bot I was able to remove all of the other delays, so a toggle might not be needed. The only delay needed is the one after writing/before reading. During my testing, there was essentially no difference in framerate after these changes. (I would recommend diffing the official one versus mine to see all of the changes.)

**Comment by chjohans** [comment]
It's great to see this level of unselfish cooperation, thanks guys!

With @EvanMulawski 's modified plugin file I can run FanControl with the Corsair Link plugin, HWiNFO with CorsairLink support enabled, and SignalRGB in parallel with no issues whatsoever. 

@Heal-Bot Hopefully these changes will make it into an upcoming build of SingalRGB. Without these changes, I can't use SignalRGB without quickly running into issues that require a reboot.

**Comment by Meh7991** [comment]
> @Heal-Bot Hopefully these changes will make it into an upcoming build of SingalRGB. Without these changes, I can't use SignalRGB without quickly running into issues that require a reboot.

Same here - this plugin now makes my H150i Pro XT sync perfectly with everything else, crash-free! Being included by default in a future SignalRGB patch would be perfect so as we don't need to go in and overwrite the original plugins every time.

**Comment by Meh7991** [comment]
Just had something strange happen. I noticed the FanControl error message was popping up every minute or so. 
I thought I would check if the H150i fans are responding in FanControl and they... Were not 😟

My log file was updating near enough every second: 
[CorsairLink.zip](https://github.com/EvanMulawski/FanControl.CorsairLink/files/13995168/CorsairLink.zip)

And this is my console log from SignalRGB: 
![image](https://github.com/EvanMulawski/FanControl.CorsairLink/assets/76539753/6a3c80e7-50a3-4efb-9e73-04ff355c8719)

The lighting is also not updating with SignalRGB. 

I haven't tampered with anything whatsoever, literally been using the same setup for the past few days since my other comment above.

**Comment by EvanMulawski** [comment]
New build. This changes the checksum failure logs to Warning but internally counts them as read failures - three read failures in a row will trigger a reboot.

**Note:** if a reboot fails, there isn't anything that can be done on the software/OS side - the device needs a power cycle (disconnecting/reconnecting the SATA power - disconnecting/reconnecting USB is not enough at that point).

[FanControl.CorsairLink-HydroPlatinumReboot-v20240121r001.zip](https://github.com/EvanMulawski/FanControl.CorsairLink/files/14002535/FanControl.CorsairLink-HydroPlatinumReboot-v20240121r001.zip)

**Comment by Meh7991** [comment]
Using the new build, after about an hour of running I heard a USB disconnect sound in Windows. I looked around my PC and saw my cooler was still updating with my RGB effect set in SignalRGB, but the centre of it was flickering in rainbow colours. 

Here is my log file: 
[CorsairLink.zip](https://github.com/EvanMulawski/FanControl.CorsairLink/files/14002758/CorsairLink.zip)

FanControl was still updating the fan speeds and all the lighting (except the centre) was working as intended - and I did not need to power cycle my system, simply exiting FC and SRGB and starting the Corsair Service fixed it. 

For now, I have reverted to the previous build (v20240112r001).

**Comment by EvanMulawski** [comment]
@Meh7991 Can you try the v20240121r001 build again but ensure the Corsair Service is stopped and disabled? Using that should not be necessary. I'm not sure why the lighting would only be partially working - if it happens again, take a screenshot of the SignalRGB logs and try just a restart of SignalRGB.

**Comment by Meh7991** [comment]
While I was using v20240121r001, I definitely had the Corsair Service disabled and stopped. I had to go into services and manually set it to "automatic" and "start" it when it happened earlier in order to fix it. 
It's definitely stopped and disabled now. 

I would have sent a screenshot of my SRGB log, but I checked it and there was nothing alarming in there - just the standard fan and pump logs. 

I will re-install the new build and set it up again early tomorrow, and report back again if it happens (or anything else happens).

**Comment by EvanMulawski** [comment]
@Meh7991 Thanks. According to your log, the device rebooted as expected after three consecutive read failures (hence the USB notification sounds). Thinking about it more, I might know why starting the Corsair Service fixed the lighting issue - I'll do some more testing on my end.

**Comment by Meh7991** [comment]
Just come across another strange interaction... 
I just downloaded a game from my Steam library, it installed perfectly, but when I launch it, after a couple of seconds FanControl pops up with that error again. 

Look at the amount of logs that happen while I have the game running: 
[CorsairLink.zip](https://github.com/EvanMulawski/FanControl.CorsairLink/files/14003068/CorsairLink.zip)
22:23:20 is when I first launched the game. Using the old plugin build (v20240112r001). 

[CorsairLink.zip](https://github.com/EvanMulawski/FanControl.CorsairLink/files/14003117/CorsairLink.zip)
This is the log with the new build (v20240121r001). 

![image](https://github.com/EvanMulawski/FanControl.CorsairLink/assets/76539753/5e09b242-7c5d-4c46-8967-acc2fa2f35ff)
And this is a screenshot of my SRGB log with the new plugin build (v20240121r001). 

I exited the game and re-launched it three times (the first attached log), and every single time is the same. 
As soon as I close the game, the logging stops. All lighting works, but the fan speeds don't update when I have game running. 
Not sure if this would be a FC problem, plugin issue, a problem with the game, or something else? 
Is there any way of narrowing the source down further? 

I am able to play any other game perfectly fine, it's literally just this one which is very strange. 

The problematic game is Crash Bandicoot N Sane Trilogy. Screw me I guess for wanting some childhood nostalgia tonight 😭

**Comment by EvanMulawski** [comment]
@Meh7991 That is very interesting. Can you perform a packet capture when able? I need to see if the game is attempting to communicate with the Corsair device.

Procedure:

1. Follow step 1 of the [USB packet capture instructions](https://github.com/EvanMulawski/FanControl.CorsairLink/wiki/USB-Packet-Capture).
2. Exit Fan Control and SignalRGB.
3. Follow step 2 of the [USB packet capture instructions](https://github.com/EvanMulawski/FanControl.CorsairLink/wiki/USB-Packet-Capture).
4. Launch the game. Wait about a minute.
5. Follow step 5 of the [USB packet capture instructions](https://github.com/EvanMulawski/FanControl.CorsairLink/wiki/USB-Packet-Capture).

**Comment by Meh7991** [comment]
@EvanMulawski Hopefully I did that capture properly! 
I exited FanControl and SignalRGB, left Corsair Service as "disabled" and it was not started. I did not open iCUE (assuming I only had to follow steps 1, 2 and 5). 

I opened the game around 10:10, played around for a bit and then quit near to 10:12. 

[CrashBandicootTrilogyCapture.zip](https://github.com/EvanMulawski/FanControl.CorsairLink/files/14008517/CrashBandicootTrilogyCapture.zip)

Let me know if I need to provide anything else. 

Side note, I re-installed the new build (v20240121r001) and the exact same thing happened after a while with the lighting. Again, nothing noteworthy in SRGB logs.

**Comment by EvanMulawski** [comment]
@Meh7991 Looks like as soon as you started the game at 10:10 it began spamming the device with requests for its Product Name.

![image](https://github.com/EvanMulawski/FanControl.CorsairLink/assets/1044791/5c878c93-d89f-4fae-a396-4f2b6d6d8690)
![image](https://github.com/EvanMulawski/FanControl.CorsairLink/assets/1044791/70e8f28c-15e5-48e4-a380-413cf816670a)

The game is requesting it continuously:

![image](https://github.com/EvanMulawski/FanControl.CorsairLink/assets/1044791/34eca13d-20a8-43f5-831a-9a78fa475e17)

This is likely causing issues reading from the device. You will have to contact the game developer about it.

**Comment by Meh7991** [comment]
Hmm is it common for games to do this, do you know? 
I don't really see it having a need to spam it endlessly - it's an offline platforming game 😂 I can maybe understand it if it's a triple-A, open-world, extreme realism game, but this?! 

Thanks for checking it for me though!

**Comment by chjohans** [comment]
I have had 2 updates for SingalRGB and I didn't repløaced the plugin with the one from @EvanMulawski, so I guess the SignalRGB devs have fixed the timeout values. It's working great now so thanks to all that contributed to this!

**Comment by Meh7991** [comment]
> I have had 2 updates for SingalRGB and I didn't repløaced the plugin with the one from @EvanMulawski, so I guess the SignalRGB devs have fixed the timeout values. It's working great now so thanks to all that contributed to this!
> 

Yeah, SignalRGB now includes the changes to the plugin that were made here, as per the [release notes](https://docs.signalrgb.com/changelogs/2-3-49#device-changes) 😁

**Comment by chjohans** [comment]
Thanks, good to know. Hope this now works as well for you all as it does for me. And please consider buying @EvanMulawski a coffee for his effort with this excellent plugin! :)

**Comment by EvanMulawski** [comment]
> Yeah, SignalRGB now includes the changes to the plugin that were made here, as per the [release notes](https://docs.signalrgb.com/changelogs/2-3-49#device-changes) 😁

Great news! I'll be posting an official 1.6 beta release soon - hopefully this weekend.

**Comment by Meh7991** [comment]
> @Meh7991 Thanks. According to your log, the device rebooted as expected after three consecutive read failures (hence the USB notification sounds). Thinking about it more, I might know why starting the Corsair Service fixed the lighting issue - I'll do some more testing on my end.

Btw, did you figure out what the cause for this was?
I tried the most recent build again yesterday and the lighting still glitched out after a while.

**Comment by EvanMulawski** [comment]
@Meh7991 Please try the attached build.

[FanControl.CorsairLink-HydroPlatinumReboot-v20240126r001.zip](https://github.com/EvanMulawski/FanControl.CorsairLink/files/14071090/FanControl.CorsairLink-HydroPlatinumReboot-v20240126r001.zip)

**Comment by Meh7991** [comment]
> @Meh7991 Please try the attached build.
> 
> [FanControl.CorsairLink-HydroPlatinumReboot-v20240126r001.zip](https://github.com/EvanMulawski/FanControl.CorsairLink/files/14071090/FanControl.CorsairLink-HydroPlatinumReboot-v20240126r001.zip)

So far, so good after 1 hour of use! 
Will keep an eye on it for the rest of the day and report back later tonight (or before if anything goes wrong).

**Comment by Meh7991** [comment]
> @EvanMulawski Hopefully I did that capture properly! I exited FanControl and SignalRGB, left Corsair Service as "disabled" and it was not started. I did not open iCUE (assuming I only had to follow steps 1, 2 and 5).
> 
> I opened the game around 10:10, played around for a bit and then quit near to 10:12.
> 
> [CrashBandicootTrilogyCapture.zip](https://github.com/EvanMulawski/FanControl.CorsairLink/files/14008517/CrashBandicootTrilogyCapture.zip)
> 
> Let me know if I need to provide anything else.
> 
> Side note, I re-installed the new build (v20240121r001) and the exact same thing happened after a while with the lighting. Again, nothing noteworthy in SRGB logs.

Just did some more testing on this... 

It seems, both FanControl and this plugin works perfectly fine when I run the game and I have SignalRGB closed. 
If I have both FC and SRGB running when I launch the game, the device bugs out. 
If I have FC closed and SRGB running when I launch the game, it still bugs out. SRGB logs contain a bunch of errors and warnings. 

Could anything further be done to the SRGB device plugin to rectify this, or would it be too much work and/or tricky to pinpoint what to change? 

I'm not asking for more changes to be added to the public release for the SRGB plugin, I'm more than happy to overwrite the file manually after every update if it means everything works for me! 😁 

Of course though, @EvanMulawski, don't go out of your way for this - if it comes to it, when I want to play it I will just settle with closing SRGB, but if it turns out to be a simple fix.. 🤩 
Although, I don't know if anything could be done as the game spams requests to the device, as you mentioned above.

**Comment by Meh7991** [comment]
@EvanMulawski, this happened after several hours of constant use.
![20240127_220924.jpg](https://github.com/EvanMulawski/FanControl.CorsairLink/assets/76539753/1d5fe99b-6511-4487-9b3e-ddb2dddd88e4)

No errors or anything, got a couple of those checksum failures in the CorsairLink log.

Restarting SRGB didn't fix it, I had to exit both FC and SRGB, start Corsair Service then stop and re-open both apps again.

**Comment by EvanMulawski** [comment]
> @EvanMulawski, this happened after several hours of constant use.
> 
> All lighting set to a dim white colour, no errors or anything, got a couple of those checksum failures in the CorsairLink log.
> 
> Restarting SRGB didn't fix it, I had to exit both FC and SRGB, start Corsair Service then stop and re-open both apps again.

Hmmm, if it happens again, can you perform a USB packet capture just before starting the Corsair Service (capture can be stopped as soon as the lighting is fixed)?

**Comment by Meh7991** [comment]
Will do - after I quit SRGB just now, the lighting changed from the blue above to flickering in all rainbow colours, working fine at the moment. 

But yeah, if/when it does it again, I'll get a capture.

**Comment by Meh7991** [comment]
I was just watching a video on YouTube, barely any apps running in the background (nothing more than usual) and I heard a USB disconnect sound. 
I saw the cooler lighting freeze and then rectify itself after about 5 seconds. Its fans' RPM in FC were slowly ramping back up from 0RPM to the value they should be set to (around 1000RPM) so it was definitely the cooler that "disconnected". 

Nothing in any logs, but I've attached my latest log file (I deleted the log after I fixed the lighting in my previous message, so this only has like 30 minutes worth of logs): 
[CorsairLink.zip](https://github.com/EvanMulawski/FanControl.CorsairLink/files/14074237/CorsairLink.zip)

It's all be working perfectly fine up until tonight! Why is this device so finnicky?!

**Comment by EvanMulawski** [comment]
> I was just watching a video on YouTube, barely any apps running in the background (nothing more than usual) and I heard a USB disconnect sound. I saw the cooler lighting freeze and then rectify itself after about 5 seconds. Its fans' RPM in FC were slowly ramping back up from 0RPM to the value they should be set to (around 1000RPM) so it was definitely the cooler that "disconnected".
> 
> Nothing in any logs, but I've attached my latest log file (I deleted the log after I fixed the lighting in my previous message, so this only has like 30 minutes worth of logs): [CorsairLink.zip](https://github.com/EvanMulawski/FanControl.CorsairLink/files/14074237/CorsairLink.zip)
> 
> It's all be working perfectly fine up until tonight! Why is this device so finnicky?!

That is "expected" behavior - the device will be instructed to reboot when three read failures occur in a row. iCUE does the same thing. It's caused by a hardware or firmware issue and the reboot is just a workaround to get it back in a good state.

**Comment by Meh7991** [comment]
Ahh yes, I remember you mentioning that a while ago.

Thought I'd mention it, unexpected USB sounds scare me lol...

I'll try and keep my PC running for most of tomorrow and if I'm able to, I'll send that USB capture.

**Comment by EvanMulawski** [comment]
Closing this as completed in v1.6.0-beta.1 as the reboot feature is done.

@Meh7991 Please open a new issue when you get that capture.
